[Fonts API] Relocate which fonts to print into wp_print_fonts()

hellofromtonya opened the pull request:
Fixes #50140 

## What?
Relocates the decision making code about which fonts to print for the iframed editor out of the script-loader and into the `wp_print_fonts()` function.

## Why?
When in the Site Editor, all registered fonts should be printed for user previewing and selection. Else, all enqueued fonts should be printed. This kind of decision making should be in the Fonts API, rather than externally in the script-loader.

## How?
This PR places the decision making logic into `wp_print_fonts()`.

### How to identify the iframed editor assets function is invoking the printing?
This function needs to know if the request to print is coming from the iframed editor for decision making about which fonts to print. This is done by passing `true` to the first parameter of `wp_prints_fonts()`. 

Notes: `true` for `$handles` is not used by `WP_Dependencies` (including in `wp_print_styles()` and `wp_print_scripts()`. Using it avoids performance hits and BC concerns with the other strategies considered (expand the next section more more details).

<details>
<summary>Strategies considered</summary>

#### Approach 1: Adding a parameter to the function's signature

Pros:
- straight-forward approach as the invoking code in the script-loader files sets it to `true`.
- avoids the overhead of firing a hook.
- avoids third-party hooking into this part of the code, i.e. a hook is available later in the call stack.
- more performant than the other approaches.

Cons:
- Deviates from `wp_print_styles()` and `wp_print_scripts()`.
- Adds a method parameter which must be maintained for BC once in WP Core.

#### Approach 2: Pass true to the first parameter as flag to indicate to populate the iframed editor handles.

Pros:
* No additional function parameter is needed, meaning no future BC concerns for adding a parameter.
* Unique flag: `true` is not used in `wp_print_styles()` or `wp_print_scripts()`.
* No performance hit.

Cons:
* Could it cause confusion? 🤔 

#### Approach 3: Add or use a hook.

##### Approach 3.1: Hook into the `'block_editor_settings_all'` to modify the `'__unstableResolvedAssets`' setting.

In WP Core, the call stack is this:

`_wp_get_iframed_editor_assets()`
invokes 
`_wp_get_iframed_editor_assets()`
which will invoke
`wp_print_fonts()`

There is a filter after this code called `'block_editor_settings_all'`.

Gutenberg uses this approach to modify the settings for the upcoming release.

Pros:
* No extra hook or function parameter to maintain for BC.

Cons:
* Once in Core, the code would be separate from `_wp_get_iframed_editor_assets()`, which would make maintaining more complex.
* Contributors would need to stay aware of changes in `_wp_get_iframed_editor_assets()` that could require changes within `wp_print_fonts()` of which setting or subsetting to modify. (Conversely if invoking directly within `_wp_get_iframed_editor_assets()` all knowledge is contained in one place for the iframed editor assets.)
* Performance hit: adding and firing a hooked callback has a performance cost.

##### Approach 3.2: Add a hook in `_wp_get_iframed_editor_assets()`.

Pros:
* Wires internal and external iframed editor asset functionality together (vs using a hook later in the call stack).

Cons:
* Adds another hook that must be maintained forever due to BC.
* Performance hit: firing a hook has a performance cost.

</details>

## TODO
- [X] Relocate from 6.2 compat file.
- [X] Relocate from 6.3 compat file.
- [X] Explore options of how to tell if the request is from the iframed editor or not.
- [x] Add tests to validate.

## Testing Instructions
### Visually Test:

Set up on your test site:
1. Activate the Gutenberg plugin.
2. Activate TT3 theme.

Before applying this PR:
1. Go to the Site Editor > Styles > Typography.
2. Using your browser's dev tools, search the HMTL for `wp-fonts-local` `<style>` element.
3. Copying the internals of that element. You'll use this to compare after applying this PR.

After applying this PR:
1. Go to the Site Editor > Styles > Typography.
2. Using your browser's dev tools, search the HMTL for `wp-fonts-local` `<style>` element.
3. Compare the styles to the before styles from above. They should match.

### Running the automated tests

All tests should pass when running the PHPUnit tests:
```
npm run test:unit:php -- --filter Tests_Fonts_WpPrintFonts
```

hellofromtonya committed:
Relocate from 6.2 script-loader

hellofromtonya committed:
Add flag to 6.3-script-loader call

hellofromtonya committed:
Add tests

hellofromtonya committed:
Assign registered to $handles.

This simplifies the code as it removes the need to modify and then reset
the `done` and `queue` properties.

hellofromtonya commented on lib/experimental/fonts-api/fonts-api.php:
```diff
@@ -176,13 +176,15 @@ function wp_register_font_provider( $name, $classname ) {
 	 *                        An empty array if none were processed.
 	 */
-	function wp_print_fonts( $handles = false ) {
+	function wp_print_fonts( $handles = false, $for_iframed_editor = false ) {
```
This function needs to know if the request to print is coming from the iframed editor for decision making about which fonts to print.

What strategies could be used?

- Approach 1: Adding a parameter to the function's signature

Pros:
- straight-forward approach as the invoking code in the script-loader files sets it to `true`.
- avoids the overhead of firing a hook.
- avoids third-party hooking into this part of the code, i.e. a hook is available later in the call stack.
- more performant than the other approaches.

Cons:
- Deviates from `wp_print_styles()` and `wp_print_scripts()`.
- Adds a method parameter which must be maintained for BC once in WP Core.

* Approach 2: Hook into the `'block_editor_settings_all'` to modify the `'__unstableResolvedAssets`' setting.

In WP Core, the call stack is this:

`_wp_get_iframed_editor_assets()`
invokes 
`_wp_get_iframed_editor_assets()`
which will invoke
`wp_print_fonts()`

There is a filter after this code called `'block_editor_settings_all'`.

Gutenberg uses this approach to modify the settings for the upcoming release.

Pros:
* No additional function parameter is needed, meaning no future BC concerns for adding a parameter.

Cons:
* Once in Core, the code would be separate from `_wp_get_iframed_editor_assets()`, which would make maintaining more complex.
* Contributors would need to stay aware of changes in `_wp_get_iframed_editor_assets()` that could require changes within `wp_print_fonts()` of which setting or subsetting to modify. (Conversely if invoking directly within `_wp_get_iframed_editor_assets()` all knowledge is contained in one place for the iframed editor assets.)
* Performance hit: adding and firing a hooked callback has a performance cost.

* Approach 3: Add a hook in `_wp_get_iframed_editor_assets()`.

Pros:
* Wires internal and external iframed editor asset functionality together (vs using a hook later in the call stack).

Cons:
* Adds another hook that must be maintained forever due to BC.
* Performance hit: firing a hook has a performance cost.

hellofromtonya committed:
Set $handles to true

hellofromtonya commented:
Now that PR #50225 has been merged into `trunk`, I'll rebase this PR on top of it.

aristath commented on lib/experimental/fonts-api/fonts-api.php:
```diff
@@ -197,6 +198,11 @@ function wp_print_fonts( $handles = false ) {
 
 		_wp_scripts_maybe_doing_it_wrong( __FUNCTION__ );
 
+		// Print all registered fonts for the iframed editor.
+		if ( true === $handles ) {
+			$handles = $registered;
+		}
+
```
Shouldn't this condition be moved _above_ the `if ( empty( $handles ) ) {` check a few lines above? 🤔

hellofromtonya replied:
Good observation @aristath! Yes, I agree with you. 

In looking further at this area of code, there's a tiny tiny micro-optimization opportunity which will benefit the front-end, i.e. when `false` then skip over these decision-making reassignment conditionals as they are unnecessary.

Changes done here https://github.com/WordPress/gutenberg/pull/50151/commits/fd7ce3332255f340ca0bfc6f9fee207224ccb1fa

hellofromtonya committed:
Combine ifs into if/elseif + micro-optimization

Combines the empty() and true checks into a if/elseif structure.

For a tiny micro-optimization, wraps these checks within a not
equal to `false`.

Why?

When using the default (i.e. on the front-end and with classic themes):

* Skips running `empty()`.
* Skips reassigning `false` to `$handle`.

These skips gain a tiny tiny performance boost.

anton-vlasenko reviewed (commented):
I've tried to follow the testing steps, and after applying the patch I couldn't find the `style#wp-fonts-local` element.
![Screenshot 2023-05-04 at 19 51 33](https://user-images.githubusercontent.com/43744263/236288089-83440af7-8215-48c6-b444-44df85599b13.png)

hellofromtonya committed:
Restore overwriting queue and done.

Printing `$handles` does not work (for some reason that needs to be investigated further). Overwriting
the queue and done properties does print and has been working since 6.2 alpha. This commit restores
the overwriting properties approach.

hellofromtonya commented:
@anton-vlasenko thank you for testing. You're right. It does not print within the Site Editor or other block editors (like Posts). 

Why?
In investigating, there's something wrong within the Fonts API or `WP_Dependencies` when injecting `$handles` into `do_items()` inside of an `ob_` circuit. Not sure yet of the root cause. 

How I fixed it:
I've restored the [overwriting of `queue` and `done` properties](https://github.com/WordPress/gutenberg/pull/50151/commits/dfd8657483d976df5829ed96adfc75a5175460e1), like it was in the 6.2-compat file. In my testing, fonts are again being printed in the Site Editor.

Restoring that logic keeps the PR's scope focused on "relocating" the logic. A follow-up PR can come later for changing the approach to copying registered fonts to `$handle` (instead of overwriting properties).

Can you re-test please @anton-vlasenko ?

anton-vlasenko reviewed (approved):
I've retested this PR after applying the changes in dfd8657483d976df5829ed96adfc75a5175460e1, and it worked in my tests. The code makes sense, so I approve this PR.

anton-vlasenko commented:
> Why? In investigating, there's something wrong within the Fonts API or `WP_Dependencies` when injecting `$handles` into `do_items()` inside of an `ob_` circuit. Not sure yet of the root cause.
> 
> How I fixed it: I've restored the [overwriting of `queue` and `done` properties](https://github.com/WordPress/gutenberg/pull/50151/commits/dfd8657483d976df5829ed96adfc75a5175460e1), like it was in the 6.2-compat file. In my testing, fonts are again being printed in the Site Editor.
> 
> Restoring that logic keeps the PR's scope focused on "relocating" the logic. A follow-up PR can come later for changing the approach to copying registered fonts to `$handle` (instead of overwriting properties).

Thank you for the thorough explanation, @hellofromtonya!

hellofromtonya commented:
Thanks for retesting and approving the PR @anton-vlasenko!

hellofromtonya commented:
Cherry-pick for minor notes:

This PR requires https://github.com/WordPress/gutenberg/pull/50225 to be cherry-picked first. Why? It uses code from PR #50225.

cc @bph

autumnfjeld commented:
cc: @arthur791004

bph commented:
I just cherry-picked this PR to the release/15.7 branch to get it included in the next release: 5682ad4240
